frontends/PyCDE/src/pycde/types.py

```diff
@@ -89,4 +91,33 @@ def dim(inner_type_or_bitwidth, *size: int, name: str = None) -> hw.ArrayType:
     ret = inner_type_or_bitwidth
   for s in size:
     ret = hw.ArrayType.get(ret, s)
-  return types.maybe_create_alias(ret, name)
+  return types.wrap(ret, name)
+
+
+# Parameterized class to subclass 'type'.
+def PyCDEType(type):
+  if type.__class__.__name__ == "_PyCDEType":
+    return type
+  type = circt.support.type_to_pytype(type)
+
+  class _PyCDEType(type.__class__):
+    """Add methods to an MLIR type class."""
+
+    # If we're subclassing a type alias, use a different 'inner' implementation.
+    if isinstance(type, hw.TypeAliasType):
+      @property
```

Comment:
Since this is going to define an `inner` method with the same signature either way, maybe sink this `if` into the `def`? It seems more natural to have one `def` with the logic inside, unless I'm missing something.

Reply:
I thought this was slick. This way, the if runs at class creation time rather than at `inner` invocation time. If you think about it, if I put the `if` inside the function I'm creating two "dynamic dispatches" -- one for the the `inner` call and one for the `if` to check the type. Putting the `if` outside the function means when `inner` gets called it's just one virtual function call!

Reply:
I'm not arguing it's slick, it's just not the most understandable Python for future maintainers (like me). I guess you are microoptimizing this code right now, so it makes sense. But please don't do this everywhere unless saving one `isinstance` call really makes a difference.

Reply:
Fair enough.

I've been obsessed with perf lately. Our internal app takes 15-20 seconds to run on a small design and there's only [one hotspot in the CIRCT python code](https://github.com/llvm/circt/blob/bcbf7288d31df56fe83ac35ffdd9ef503cdb052c/lib/Bindings/Python/circt/support.py#L79) followed by a long tail of functions -- many of which are tiny functions like this -- so I've been optimizing any new function which may be called a bunch. Though micro-optimizations like these may just be silly in Python -- it may be more efficient to move more logic into C++.